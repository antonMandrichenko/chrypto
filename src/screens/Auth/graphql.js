import gql from "graphql-tag";

export const SIGN_UP = gql`
  mutation signUp(
    $username: String!
    $password: String
    $promoCode: String
    $captchaResponse: String
  ) {
    signUp(
      username: $username
      password: $password
      promoCode: $promoCode
      captchaResponse: $captchaResponse
    ) {
      user {
        id
        username
        email
        affiliateCode
      }
      token
    }
  }
`;

export const LOGIN = gql`
  mutation login($username: String!, $password: String!) {
    login(username: $username, password: $password) {
      user {
        id
        username
        email
        affiliateCode
      }
      token
    }
  }
`;