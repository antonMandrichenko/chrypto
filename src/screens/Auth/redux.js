import { userActions } from "../../store/actions/auth.actions";
import { fetchActions } from "../../store/actions/fetch.actions";

export const mapStateToProps = state => {
  return {
    isLoading: state.fetch.isLoading,
    errorLoading: state.fetch.error
  };
};

export const mapDispatchToProps = (dispatch, ownProps) => {
  return {
    loginSuccess: data =>
      dispatch(userActions.loginSuccess(ownProps, dispatch, data)),
    loginFailure: data =>
      dispatch(userActions.loginFailure(ownProps, dispatch, data)),
    registrationSuccess: data =>
      dispatch(userActions.registrationSuccess(ownProps, data)),
    registrationFailure: err => dispatch(userActions.registrationFailure(err)),
    fetchBegin: () => dispatch(fetchActions.fetchBegin()),
    fetchSuccess: () => dispatch(fetchActions.fetchSuccess()),
    fetchFailure: err => dispatch(fetchActions.fetchFailure(err)),
    noResponse: () => dispatch(fetchActions.noResponse()),
    noFetch: () => dispatch(fetchActions.noFetch())
  };
};
