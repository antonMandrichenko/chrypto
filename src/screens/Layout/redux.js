// import { userActions } from "../../store/actions/auth.actions";

export const mapStateToProps = state => {
  return {
    id: state.registration.id || state.authentication.id,
  };
};
