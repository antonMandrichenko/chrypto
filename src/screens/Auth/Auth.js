import React from "react";
import PropTypes from "prop-types";
import { makeStyles, useTheme } from "@material-ui/core/styles";
import Paper from "@material-ui/core/Paper";
import Grid from "@material-ui/core/Grid";
import AppButton from "../../components/Button";
import AppInput from "../../components/Input";
import Tabs from "@material-ui/core/Tabs";
import Tab from "@material-ui/core/Tab";
import Typography from "@material-ui/core/Typography";
import Box from "@material-ui/core/Box";
import { connect } from "react-redux";
import { compose } from "redux";
import { withRouter } from "react-router-dom";
import { mapStateToProps, mapDispatchToProps } from "./redux";
import { SIGN_UP, LOGIN } from "./graphql";
import { useMutation } from "@apollo/react-hooks";

const propTypes = {};

function TabPanel(props) {
  const { children, value, index, ...other } = props;

  return (
    <Typography
      component="div"
      role="tabpanel"
      hidden={value !== index}
      id={`full-width-tabpanel-${index}`}
      aria-labelledby={`full-width-tab-${index}`}
      {...other}
    >
      <Box pt={3} pb={3}>
        {children}
      </Box>
    </Typography>
  );
}

TabPanel.propTypes = {
  children: PropTypes.node,
  index: PropTypes.any.isRequired,
  value: PropTypes.any.isRequired
};

function a11yProps(index) {
  return {
    id: `full-width-tab-${index}`,
    "aria-controls": `full-width-tabpanel-${index}`
  };
}

const useStyles = makeStyles(theme => ({
  root: {
    width: "100vw",
    height: "100vh",
    backgroundColor: theme.palette.background.default
  },
  paper: {
    width: "40vw",
    height: "auto",
    padding: theme.spacing(4),
    borderRadius: 10,
    border: "1px solid #2b3044",
    [theme.breakpoints.down("sm")]: {
      width: "50vw"
    },
    [theme.breakpoints.down("xs")]: {
      width: "80vw"
    },
    [theme.breakpoints.up("lg")]: {
      width: "30vw"
    }
  },
  colorSecondary: {
    background: "linear-gradient(131deg, #da1b60 0%, #ff8a00 100%)"
  },
  tabs: {
    backgroundColor: theme.palette.background.paper,
    width: "100%"
  },
  "Mui-selected": {
    color: "#fff !important"
  }
}));

function Auth(props) {
  const {
    loginSuccess,
    registrationSuccess,
    registrationFailure,
    loginFailure,
    fetchBegin,
    fetchSuccess,
    fetchFailure,
    noResponse,
    noFetch,
    isLoading,
    errorLoading
  } = props;
  const classes = useStyles();
  const theme = useTheme();
  const [value, setValue] = React.useState(0);
  const [inputValues, setInputValues] = React.useState({
    "email-login": "",
    "password-login": "",
    "email-register": "",
    "phone-register": "",
    "password-register": "",
    "confirm password-register": ""
  });

  const [signUp] = useMutation(SIGN_UP);
  const [login] = useMutation(LOGIN);

  const handleChange = name => event => {
    console.log(name, event.target.value);
    errorLoading && noFetch();
    setInputValues({ ...inputValues, [name]: event.target.value });
  };

  const handleChangeValue = (event, newValue) => {
    errorLoading && noFetch();
    setValue(newValue);
  };

  const validatedForm = (email, password, confirmPassword, phone) => {
    const isValidEmail = /^[a-zA-Z0-9]+@[a-zA-Z0-9]+\.[A-Za-z]+$/.test(email);
    const isValidPassword = password.length > 6;
    const isPasswordConfirm = confirmPassword
      ? password === confirmPassword
      : null;
    const isValidPhone = phone
      ? /\(?([0-9]{3})\)?([ .-]?)([0-9]{3})\2([0-9]{4})/.test(phone)
      : null;
    if (!isValidEmail) {
      fetchFailure("Not valid email");
      return false;
    }
    if (!isValidPassword) {
      fetchFailure("Not valid password. Password must have more 6 symbols");
      return false;
    }
    if (confirmPassword || confirmPassword === "" || phone || phone === "") {
      if (!isValidPhone) {
        fetchFailure("Not valid phone number");
        return false;
      }
      if (!isPasswordConfirm) {
        fetchFailure("Not confirm password");
        return false;
      }
    }
    return true;
  };

  const signIn = async e => {
    e.preventDefault();

    console.log("begin");
    const email = inputValues["email-login"];
    const password = inputValues["password-login"];
    if (validatedForm(email, password)) {
      try {
        fetchBegin();
        const loginData = await login({
          variables: {
            username: email,
            password
          }
        });
        if (!loginData) {
          noResponse();
        } else {
          fetchSuccess();
          loginSuccess(loginData);
        }
      } catch (e) {
        fetchFailure(e.message);
      }
    }
  };

  const register = async e => {
    e.preventDefault();
    const email = inputValues["email-register"];
    const password = inputValues["password-register"];
    const confirmPassword = inputValues["confirm password-register"];
    const phoneNumber = inputValues["phone-register"];
    if (validatedForm(email, password, confirmPassword, phoneNumber)) {
      try {
        fetchBegin();
        const registerData = await signUp({
          variables: {
            username: inputValues["email-register"],
            password: inputValues["password-register"]
          }
        });
        if (!registerData) {
          noResponse();
        } else {
          fetchSuccess();
          registrationSuccess(registerData);
        }
      } catch (e) {
        fetchFailure(e.message);
      }
    }
  };

  return (
    <Grid
      container
      justify="center"
      alignItems="center"
      className={classes.root}
    >
      <Paper className={classes.paper}>
        <h6
          style={{
            textAlign: "center",
            color: "#fff",
            fontSize: 15
          }}
        >
          SETTINGS
        </h6>
        <div className={classes.tabs}>
          <Tabs
            value={value}
            onChange={handleChangeValue}
            textColor="secondary"
            variant="fullWidth"
          >
            <Tab
              label="LOGIN"
              {...a11yProps(0)}
              className={classes["Mui-selected"]}
              disabled={isLoading}
            />
            <Tab
              label="REGISTER"
              {...a11yProps(1)}
              className={classes["Mui-selected"]}
              disabled={isLoading}
            />
          </Tabs>
          <TabPanel value={value} index={0} dir={theme.direction}>
            <form noValidate onSubmit={signIn}>
              <AppInput
                name="email"
                type="email"
                handleChange={handleChange}
                layout="login"
                isLoading={isLoading}
              />
              <AppInput
                name="password"
                type="password"
                handleChange={handleChange}
                layout="login"
                isLoading={isLoading}
              />
              <AppButton isLoading={isLoading}>Sign in</AppButton>
            </form>
          </TabPanel>
          <TabPanel value={value} index={1} dir={theme.direction}>
            <form noValidate onSubmit={register}>
              <AppInput
                name="email"
                type="email"
                handleChange={handleChange}
                layout="register"
                isLoading={isLoading}
              />
              <AppInput
                name="phone"
                type="phone"
                handleChange={handleChange}
                layout="register"
                isLoading={isLoading}
              />
              <AppInput
                name="password"
                type="password"
                handleChange={handleChange}
                layout="register"
                isLoading={isLoading}
              />
              <AppInput
                name="confirm password"
                type="password"
                handleChange={handleChange}
                layout="register"
                isLoading={isLoading}
              />
              <AppButton isLoading={isLoading}>Register</AppButton>
            </form>
          </TabPanel>
          {errorLoading !== "" && (
            <Typography color="error">
              {typeof errorLoading === "string"
                ? errorLoading.replace("GraphQL error:", "")
                : errorLoading}
            </Typography>
          )}
          <h6
            className="text-center"
            style={{
              textAlign: "center",
              color: "#737375",
              fontSize: 14,
              fontFamily: "FuturaPTDemi",
              marginTop: 18,
              marginBottom: 0
            }}
          >
            Forgot Password
          </h6>
        </div>
      </Paper>
    </Grid>
  );
}

Auth.propTypes = propTypes;

export default compose(
  withRouter,
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)(Auth);
