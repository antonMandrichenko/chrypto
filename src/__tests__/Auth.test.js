import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import Auth from "../screens/Auth";
import { createStore } from "redux";
import { Provider } from "react-redux";
import rootReducer from "../store/reducers";
import { MockedProvider } from "@apollo/react-testing";
import { MemoryRouter } from "react-router-dom";
import { SIGN_UP } from "../screens/Auth/graphql";

const initialState = {};

const mocks = [
  {
    request: {
      query: SIGN_UP,
      variables: {
        name: "Buck"
      }
    },
    result: {
      data: {
        dog: { id: "1", name: "Buck", breed: "bulldog" }
      }
    }
  }
];

function renderWithReduxApollo(
  ui,
  { initialState, store = createStore(rootReducer, initialState) } = {}
) {
  return {
    ...render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Provider store={store}>{ui}</Provider>
      </MockedProvider>
    ),
    store
  };
}

it("loads and displays auth form", async () => {
  const { asFragment } = renderWithReduxApollo(
    <MemoryRouter>
      <Auth name="Buck" />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
