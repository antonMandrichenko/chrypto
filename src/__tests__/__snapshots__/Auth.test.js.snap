// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads and displays auth form 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root makeStyles-root-1 MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation1 makeStyles-paper-2 MuiPaper-rounded"
    >
      <h6
        style="text-align: center; color: rgb(255, 255, 255); font-size: 15px;"
      >
        SETTINGS
      </h6>
      <div
        class="makeStyles-tabs-4"
      >
        <div
          class="MuiTabs-root"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed"
            style="overflow: hidden;"
          >
            <div
              class="MuiTabs-flexContainer"
              role="tablist"
            >
              <button
                aria-controls="full-width-tabpanel-0"
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary makeStyles-Mui-selected-5 Mui-selected MuiTab-fullWidth"
                id="full-width-tab-0"
                role="tab"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  LOGIN
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-controls="full-width-tabpanel-1"
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary makeStyles-Mui-selected-5 MuiTab-fullWidth"
                id="full-width-tab-1"
                role="tab"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  REGISTER
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="PrivateTabIndicator-root-160 PrivateTabIndicator-colorSecondary-162 MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
        <div
          aria-labelledby="full-width-tab-0"
          class="MuiTypography-root MuiTypography-body1"
          dir="ltr"
          id="full-width-tabpanel-0"
          role="tabpanel"
        >
          <div
            class="MuiBox-root MuiBox-root-195"
          >
            <form
              novalidate=""
            >
              <div
                class="form-group"
              >
                <label
                  class="mylabel"
                  for="email-login"
                >
                  EMAIL
                </label>
                <input
                  class="form-control form-control-lg myinputclass"
                  id="email-login"
                  type="email"
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  class="mylabel"
                  for="password-login"
                >
                  PASSWORD
                </label>
                <input
                  class="form-control form-control-lg myinputclass"
                  id="password-login"
                  type="password"
                />
              </div>
              <button
                class="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-198 MuiButton-text"
                tabindex="0"
                type="submit"
              >
                <span
                  class="MuiButton-label WithStyles(ForwardRef(Button))-label-199"
                >
                  SIGN IN
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </form>
          </div>
        </div>
        <div
          aria-labelledby="full-width-tab-1"
          class="MuiTypography-root MuiTypography-body1"
          dir="ltr"
          hidden=""
          id="full-width-tabpanel-1"
          role="tabpanel"
        >
          <div
            class="MuiBox-root MuiBox-root-228"
          >
            <form
              novalidate=""
            >
              <div
                class="form-group"
              >
                <label
                  class="mylabel"
                  for="email-register"
                >
                  EMAIL
                </label>
                <input
                  class="form-control form-control-lg myinputclass"
                  id="email-register"
                  type="email"
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  class="mylabel"
                  for="phone-register"
                >
                  PHONE
                </label>
                <input
                  class="form-control form-control-lg myinputclass"
                  id="phone-register"
                  type="phone"
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  class="mylabel"
                  for="password-register"
                >
                  PASSWORD
                </label>
                <input
                  class="form-control form-control-lg myinputclass"
                  id="password-register"
                  type="password"
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  class="mylabel"
                  for="confirm password-register"
                >
                  CONFIRM PASSWORD
                </label>
                <input
                  class="form-control form-control-lg myinputclass"
                  id="confirm password-register"
                  type="password"
                />
              </div>
              <button
                class="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-198 MuiButton-text"
                tabindex="0"
                type="submit"
              >
                <span
                  class="MuiButton-label WithStyles(ForwardRef(Button))-label-199"
                >
                  REGISTER
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </form>
          </div>
        </div>
        <h6
          class="text-center"
          style="text-align: center; color: rgb(115, 115, 117); font-size: 14px; font-family: FuturaPTDemi; margin-top: 18px; margin-bottom: 0px;"
        >
          Forgot Password
        </h6>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
